### PR TITLE
Use generic writer to write MPS files, criterion, interco, etc.

### PR DESCRIPTION
**Pull request by flomnes**
- Remove `Application::pResultWriter`
- Add `Study::getWriter`
- In MPS-writing functions, replace `fprintf` by `IResultWriter::addJob` from a buffer.
- In function `Study::createFileIntoOutputWithExtension`, do not check for the existence of files. Instead, pass the resolution number (1st or 2nd resolution) to ensure that file names are unique, removing the need for this check.

### CODE REVIEW

File: src/libs/antares/study/mps.cpp [hunk 1]
```diff
@@ -39,43 +39,25 @@ namespace Data
 {
 std::string Study::createFileIntoOutputWithExtension(const YString& prefix,
                                                      const YString& extension,
-                                                     uint numSpace) const
+                                                     uint numSpace,
+                                                     int n) const
```
Comment (guilpier-code) [inline src/libs/antares/study/mps.cpp]: What is argument **n** ? We should give it a more explicit name. As it has a default value (=0) and is never given another value, why appending a "-0" in the file name.

File: src/libs/antares/study/mps.cpp [hunk 2]
```diff
@@ -39,43 +39,25 @@ namespace Data
 {
 std::string Study::createFileIntoOutputWithExtension(const YString& prefix,
                                                      const YString& extension,
-                                                     uint numSpace) const
+                                                     uint numSpace,
```
Comment (guilpier-code) [inline src/libs/antares/study/mps.cpp]: The function name is not related to what it does, we should rename it **createWeeklySimulationPath**.
The source file hasn't got a good name as well.

File: src/solver/utils/mps_utils.cpp [hunk 3]
```diff
@@ -19,9 +19,9 @@ constexpr size_t OPT_APPEL_SOLVEUR_BUFFER_SIZE = 256;
 
 static void printHeader(Clob& Flot, int NombreDeVariables, int NombreDeContraintes)
```
Comment (guilpier-code) [inline src/solver/utils/mps_utils.cpp]: **Flot** does mean anything to me. Why not **buffer** or **memoire_tampon** (french)

File: src/solver/utils/mps_utils.cpp [hunk 4]
```diff
@@ -246,14 +240,11 @@ void OPT_dump_spx_fixed_part(const PROBLEME_SIMPLEXE* Pb, uint numSpace)
...
-    // auto archive = study->pZipArchive;
-    // archive->addData(filename,
-    //                  Flot.c_str(),
-    //                  Flot.size());
-
-    // archive->close();
-    // archive->open(libzippp::ZipArchive::Write);
+    auto study = Data::Study::Current::Get();
+    const auto filename
+      = study->createFileIntoOutputWithExtension("problem-fixed-part", "mps", numSpace);
```
Comment (guilpier-code) [inline src/solver/utils/mps_utils.cpp]: As I said in **mps.cpp** review, prepending output folder was removed from the file path

File: src/solver/optimisation/opt_export_structure.cpp [hunk 5]
```diff
@@ -66,14 +66,39 @@ void OPT_ExportInterco(const Antares::Data::Study& study,
                        PROBLEME_HEBDO* ProblemeHebdo,
                        uint numSpace)
 {
-  // TODO
+    // Interco are exported only once for first year
+    if (ProblemeHebdo->firstWeekOfSimulation)
+    {
+        Yuni::Clob Flot;
```
Comment (guilpier-code) [inline src/solver/optimisation/opt_export_structure.cpp]: **Flot** ==> **buffer** ? **memoire_tampon** ?

File: src/solver/simulation/solver.hxx [hunk 6]
```diff
@@ -281,6 +281,8 @@ inline ISimulation<Impl>::ISimulation(Data::Study& study,
         pYearByYear = false;
 
     pHydroHotStart = (study.parameters.initialReservoirLevels.iniLevels == Data::irlHotStart);
+
+    study.setWriter(pResultWriter);
```
Comment (guilpier-code) [inline src/solver/simulation/solver.hxx]: As said earlier review made on the "zip" matter, the **study->parameters** already has the info to create a writer, we should create it in the study after reading this info, and pass this writer to component that need it.

File: src/solver/utils/mps_utils.cpp [hunk 7]
```diff
@@ -290,17 +276,14 @@ void OPT_dump_spx_variable_part(const PROBLEME_SIMPLEXE* Pb, uint numSpace)
...
-    // archive->open(libzippp::ZipArchive::Write);
+    auto study = Data::Study::Current::Get();
+    const auto filename
+      = study->createFileIntoOutputWithExtension("problem-variable-part", "mps", numSpace);
+    auto writer = study->getWriter();
+    writer->addJob(filename, Flot);
 }
 
-void OPT_EcrireJeuDeDonneesLineaireAuFormatMPS(void* Prob, uint numSpace)
+void OPT_EcrireJeuDeDonneesLineaireAuFormatMPS(void* Prob, uint numSpace, int n)
```
Comment (guilpier-code) [inline src/solver/utils/mps_utils.cpp]: Parameter **n** is unused (and should be renamed [**optimNumber** ?])

File: src/libs/antares/study/mps.cpp [hunk 8]
```diff
@@ -39,43 +39,25 @@ namespace Data
...
+                                                     int n) const
 {
-    static std::map<YString, int> count;
-
     // Empty log entry
     logs.info();
 
-    // Preparing the filename where to write the MPS file
```
Comment (guilpier-code) [inline src/libs/antares/study/mps.cpp]: As this function does not need to get the output folder from the study, it's not related to the study anymore and should not be a method of the study : it belongs more to the place it is used : for instance **mps_utils.h/cpp**

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/utils/mps_utils.cpp (hunk 7 above)
Comment (flomnes) [reply src/solver/utils/mps_utils.cpp]: It's now used

File: src/solver/simulation/solver.hxx (hunk 6 above)
Comment (flomnes) [reply src/solver/simulation/solver.hxx]: Will do it elsewhere

File: src/libs/antares/study/mps.cpp (hunk 1 above)
Comment (flomnes) [reply src/libs/antares/study/mps.cpp]: name was changed, more expressive

File: src/libs/antares/study/mps.cpp (hunk 2 above)
Comment (flomnes) [reply src/libs/antares/study/mps.cpp]: Renamed into `getFilenameWithExtension`.

**Comment by flomnes** [comment]
@guilpier-code DONE

File: src/libs/antares/study/mps.cpp (hunk 1 above)
Comment (flomnes) [reply src/libs/antares/study/mps.cpp]: This will be obsolete with the MPS writer interface

**Review by guilpier-code — APPROVED** [review]

**Pull request merged**